Sources/SwiftSyntax/BumpPtrAllocator.swift

```diff
@@ -0,0 +1,113 @@
+//===---------- BumpPtrAllocator.swift - Bump-pointer Allocaiton ----------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2022 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+/// Bump-pointer allocation.
+class BumpPtrAllocator {
+  typealias Slab = UnsafeMutableRawBufferPointer
+
+  static private var SLAB_SIZE: Int = 4096
+  static private var GLOWTH_DELAY: Int = 128
+  static private var SLAB_ALIGNMENT: Int = 8
+
+  private var slabs: [Slab]
+  private var currentPtr: UnsafeMutableRawPointer?
+  private var endPtr: UnsafeMutableRawPointer?
+  private var customSizeSlabs: [Slab]
+  private(set) var totalSizeAllocated: Int
```

Comment:
Technically, this is only an underestimation of the bump allocator’s memory footprint because if there are bytes in a slab that were (not yet) used, they don’t contribute to this count, right? Maybe it’s worth pointing out that distinction in a comment.

Reply:
Yeah, you are right. We probably want two metrics like:
- `totalMemorySizeAllocatedFromSystem`: `llvm::BumpPtrAllocator.getTotalMemory()`
- `totalMemorySizeClaimedByClients`: `llvm::BumpPtrAllocator.getBytesAllocated()`
Not sure about the naming though. Any suggestions?

Reply:
Ended up with mildly following `llvm::BumpPtrAllocator`.
`totalMemorySize` / `totalByteSizeAllocated`